### DISCUSSION TITLE
Problem with Stream response

### DISCUSSION THREAD

**Discussion by danechitoaie** [litestar-org/litestar#1622] [Q&A]
I have the following controller 
```python
    @post()
    async def get_chat(
        self,
        data: ChatRequest,
        prompts: Prompts,
        httpx_client: AsyncClient,
    ) -> Stream:
        ...
        return Stream(iterator=httpx_stream(httpx_client, httpx_json))
```

and my `httpx_stream` function is:

```python
async def httpx_stream(httpx_client: AsyncClient, httpx_json: dict) -> AsyncGenerator[bytes, None]:
    async with aconnect_sse(
        httpx_client,
        "POST",
        "https://api.openai.com/v1/chat/completions",
        json=httpx_json,
    ) as event_source:
        async for event in event_source.aiter_sse():
            if event.data == "[DONE]":
                break
            data = decode_json(event.data)
            for choice in data["choices"]:
                delta = choice["delta"]
                if "content" in delta:
                    yield encode_json(
                        {
                            "status": "OK",
                            "text": delta["content"],
                        }
                    )
```

And the problem is that this works the right way about 90% of the time.
So most of the requests are coming like:
```
{"status":"OK","text":"Hello"}
{"status":"OK","text":"!"}
{"status":"OK","text":" How"}
{"status":"OK","text":" can"}
{"status":"OK","text":" I"}
```

so one JSON object per HTTP chunk response (per yield).

But randomly it also comes like this:
```
{"status":"OK","text":"Hello"}
{"status":"OK","text":"!"}
{"status":"OK","text":" How"}{"status":"OK","text":" can"}
{"status":"OK","text":" I"}
```

So it seems sometimes it puts 2 yields on same HTTP chunk.
Any idea what's going on? Why it doesn't always respect one yield per chunk?

**Comment by provinzkraut** [comment]
Can you provide an MCVE of this without the httpx and SSE stuff?

**Reply by danechitoaie** [reply]
```python
from typing import AsyncGenerator
from starlite import Starlite, Controller, post, Stream
from starlite.utils.serialization import encode_json


async def make_stream() -> AsyncGenerator[bytes, None]:
    while True:
        yield encode_json({"hello": "world"})


class MyController(Controller):
    path = "/stream"

    @post()
    async def stream(self) -> Stream:
        return Stream(iterator=make_stream())


app = Starlite(route_handlers=[MyController])
```

```
uvicorn my_app.app:app --loop uvloop --no-access-log
```

And the Node.js client reading the stream:

```javascript
const { fetch } = require("undici");

async function main() {
	const res = await fetch(
		"http://127.0.0.1:8000/stream",
		{
			method: "POST",
			body: "",
		},
	);

	const decoder = new TextDecoder("utf-8");
	for await (const chunk of res.body) {
		const data = decoder.decode(chunk);
		console.log(data);
	}
}

main();
```

This should output one `{"hello":"world"}` per line (each line being a http chunk) but instead it groups many in same chunk.
I guess maybe this has something to to with some kind of buffering that's happening somewhere on the server?

This seems to be happening when the code yields too fast. If I add a `await asyncio.sleep(1)` after yield then they get outputted correctly one on each line (so one per chunk).

Is there any way to flush the output, or to disable this buffering? I don't want to `sleep` and artificially delay the response as a fix.

**Reply by provinzkraut** [reply]
Are you sure that doesn't have to do anything with your consumer side?

I tried reproducing this with httpx, and letting it run for 100,000 iterations didn't cause any errors.

What I tried:

- uvicorn without uvloop
- uvicorn with uvloop
- httpx iterating synchronously
- httpx iterating asynchronously

Can you share versions of Starlite and uvicorn?

**Reply by provinzkraut** [reply]
Okay, disregard the previous answer. I was able to reproduce it. I am going out on a limb here and say it's caused by uvicorn, since I was able to reproduce the same behaviour with Starlette, Blacksheep and Sanic.

I would also say that this is not a bug. As far as I am aware, the ASGI spec does not make any guarantees about the chunking behaviour, and its left up to the server to decide how to serve them. The relevant section only specifies that servers must flush their send buffer before returning from a `send` call, but does not make any assumption about how frequently flushes should occur in between. https://asgi.readthedocs.io/en/latest/specs/www.html?highlight=stream#response-body-send-event

> Continues sending a response to the client. Protocol servers must flush any data passed to them into the send buffer before returning from a send call. If more_body is set to False, and the server is not expecting Response Trailers this will close the connection.

The proper way to handle this would be to in your client not make the assumption that each chunk represents exactly one frame.

**Reply by danechitoaie** [reply]
Ok, then I have to find another way to handle streaming lines of JSON.

**Reply by danechitoaie** [reply]
But, this would mean that each yield would have to be flushed no? Am I understanding this wrong?
> Protocol servers must flush any data passed to them into the send buffer before returning from a send call

And Litestar is doing an `await send()` per chunk: https://github.com/litestar-org/litestar/blob/main/litestar/response/streaming.py#L104-L110

```python
        async for chunk in self.iterator:
            stream_event: HTTPResponseBodyEvent = {
                "type": "http.response.body",
                "body": chunk if isinstance(chunk, bytes) else chunk.encode(self.encoding),
                "more_body": True,
            }
            await send(stream_event)
```

So it should flush each chunk.

**Reply by provinzkraut** [reply]
I'd have to look at some uvicorn source code to be sure, but the way I understand this 

> Protocol servers must flush any data passed to them into the send buffer before returning from a send call

Is that each `await send(event)` must flush data into the server's buffer, but does not require the server to flush its buffer or actually perform a send of the data therein.

Reading the uvicorn documentation, [this section](https://www.uvicorn.org/server-behavior/#write-flow-control) seems to confirm this:

> If the write buffer passes a high water mark, then Uvicorn ensures the ASGI send messages will only return once the write buffer has been drained below the low water mark.

`send` calls first write to the buffer, but there are no guarantees as to how and when uvicorn will actually send the contents of that buffer during the lifetime of a request.

**Reply by danechitoaie** [reply]
Then I'll need to find another way to encode and send the data in chunks then in a way that can be correctly parsed on the client side.

**Reply by provinzkraut** [reply]
I think sending JSON in chunks like that isn't a great idea anyway, since it doesn't lend itself that well to being chunked. 

Maybe you can use WebSockets or SSE?

**Reply by danechitoaie** [reply]
SSE was my first option but Litestar does not support it.

**Reply by provinzkraut** [reply]
We have an issue to track SSE support: #1578.

PRs are welcome (=
